### DISCUSSION TITLE
Using without the server — is there a use-case of hocuspocus frontend components, without a server?

### DISCUSSION THREAD

**Discussion by floer32** [ueberdosis/hocuspocus#156] [General]
I have a browser extension use case that's not collaborative editing (at first), but rather, supporting a single user with multiple tabs in one browser. Got curious if hocuspocus would be of use here — and if so, would it only be client-side usage? (Can synchronization features work with just broadcastchannel/etc, and no server? For a single user in various tabs.)

~So I was reviewing the source, and thinking about if/how hocuspocus could be used in a browser extension. A hocuspocus provider could use the browser API to exchange messages with the background-page (or in mv3, calling service worker), rather than a server. Provider would communicate with a browser extension background process, via browser API - [messages](https://www.npmjs.com/package/@extend-chrome/messages) &/or [alarms](https://stackoverflow.com/questions/66391018/how-do-i-call-a-function-periodically-in-a-manifest-v3-chrome-extension)).~

~Anybody have a sense of feasibility?~

Update a while later:

* Yes Yjs could make sense in a browser extension
* YMMV re: using hocuspocus without collaboration features needing a server. It might just be an impedance mismatch, since hocuspocus is designed for collaboration.

**Comment by floer32** [comment]
Maybe this is misguided. If it seems misguided, let me know? 😅  @tommoor

**Reply by hanspagel** [reply]
I don't know how we can help here. Using Y.js really depends on your use case and we can’t give a recommendation there, but I don’t see how this is related to the Node.js backend hocuspocus at all. 🤔

**Reply by floer32** [reply]
Yeah I figured, it's a novel use case and I'm probably trying the wrong abstractions. (For posterity: currently I'm trying out using broadcast-channel (+ [nanostores](https://github.com/nanostores/nanostores)), and plan to use hocuspocus later for true collaborative editing features)

**Reply by meatflavourdev** [reply]
Head over to [https://discuss.yjs.dev/](https://discuss.yjs.dev/) -- that's the appropriate place for a question like this.

But since we're here... I’m considering using Yjs for browser extension… the browser extension background.js has listeners set for messages from content tabs and from the chrome APIs. It writes them in an event store append-only array which is broken up semi-regularly into subdocuments, so the entire history only has to be loaded into local memory as necessary or to the last state snapshot (using event sourcing for this particular project)

The NodeJS server, hocuspocus, in this case, take these events and passes them to a service broker which takes care of the transport to the message broker and to the persistence service, analytics, whatever. The hocuspocus/gateway service is a producer in this case. It is also a consumer of some aggregation services or whatever is needed and this is automatically passed back down from the service broker’s reception, into a subdocument corresponding to the particular service to which we are subscribed. Handy eh?

I’ve essentially replaced a APIful gateway service with an API-less Yjs streaming service that sucks up my events and hands me back down aggregates or additional event log entries (as updates) as needed. I don’t have to do anything but push up events and react to the changes in state. I think this is a good idea. Haven’t tested it yet.

<img width="559" alt="Screenshot 2021-08-24 041725" src="https://user-images.githubusercontent.com/13112048/130607593-40ea8682-a0d1-410e-9ff5-58766231b094.png">

(Obv, reception of new events would trigger an update to the state which is only a projection of the event log-- notably, the state is not sent to the backend, only projection snapshots of state corresponding to a particular master event IDs are received from the snapshot service as a sanity check. When we make or receive a new snapshot, we end the event log at that ID and start a new subdocument such that each snapshot is corresponds with each eventlog subdocument)

I wonder if YJjs would be a good fit for an entire microservices backend? I mean… why stop at the frontend? :thinking: I cannot say, as I am most certainly not planning on implementing it. :stuck_out_tongue_winking_eye: Perhaps it’d be some kind of distributed stateful orchestration microservices type deal, like https://cloudstate.io/(?)
